File: thinking-categorically/natural-transformation/list-length/list-length.hs
-- The List functor
-- fmap :: (a -> b) -> [a] -> [b]

-- Natural transformation from List to Const Int
listLength :: [a] -> Int
listLength = length

-- Verification of naturality:
-- For any function f :: a -> b and list xs :: [a]:
-- listLength (fmap f xs) == listLength xs
-- This holds because length doesn't depend on the elements

-- Example usage:
example1 :: Bool
example1 = let f = (*2) :: Int -> Int
               xs = [1, 2, 3] :: [Int]
           in listLength (map f xs) == listLength xs  -- True

main :: IO ()
main = print example1

File: thinking-categorically/natural-transformation/list-length/list-length.ts
// The Array functor (built into JavaScript/TypeScript)
// Array.map :: (a -> b) -> Array<a> -> Array<b>

// Natural transformation from Array to number
const arrayLength = <A>(arr: Array<A>): number => arr.length;

// Verification of naturality:
// For any function f: A -> B and array arr: Array<A>:
// arrayLength(arr.map(f)) === arrayLength(arr)

// Example usage:
const f = (x: number) => x * 2;
const arr = [1, 2, 3];
const natural = arrayLength(arr.map(f)) === arrayLength(arr); // true

console.log(`Naturality check: ${natural}`);